File: node-src/domain/import-moves.ts
import { ProcessFailure } from "./errors.js";

const RESOURCE_TYPE = /^[a-z][a-z0-9_]*$/;
const MAX_GENERATED_IMPORT_PAIRS = 50_000;
const MAX_IMPORT_MOVE_CANDIDATES = 50_000;

export interface GeneratedImportPair {
  readonly key: string;
  readonly importId: string;
}

export interface ImportMove {
  readonly oldKey: string;
  readonly newKey: string;
}

export type ImportMoveSuppressionReason =
  | "ambiguous"
  | "duplicate_from"
  | "key_swap"
  | "destination_occupied";

export interface ImportMoveSuppression extends ImportMove {
  readonly importId: string;
  readonly reason: ImportMoveSuppressionReason;
}

export interface ImportMoveDerivation {
  readonly moves: readonly ImportMove[];
  readonly suppressed: readonly ImportMoveSuppression[];
}

export interface ParsedHclQuotedString {
  readonly value: string;
  readonly end: number;
}

function fail(code: string, message: string): never {
  throw new ProcessFailure({
    code,
    category: "domain",
    message,
  });
}

function requireResourceType(resourceType: string): void {
  if (typeof resourceType !== "string" || !RESOURCE_TYPE.test(resourceType)) {
    fail(
      "INVALID_IMPORT_RESOURCE_TYPE",
      "import resource type must be a canonical Terraform identifier",
    );
  }
}

function compareCodePoints(left: string, right: string): number {
  let leftIndex = 0;
  let rightIndex = 0;
  while (leftIndex < left.length && rightIndex < right.length) {
    const leftPoint = left.codePointAt(leftIndex) ?? 0;
    const rightPoint = right.codePointAt(rightIndex) ?? 0;
    if (leftPoint !== rightPoint) {
      return leftPoint - rightPoint;
    }
    leftIndex += leftPoint > 0xffff ? 2 : 1;
    rightIndex += rightPoint > 0xffff ? 2 : 1;
  }
  return (leftIndex < left.length ? 1 : 0)
    - (rightIndex < right.length ? 1 : 0);
}

function comparePairs(
  left: GeneratedImportPair,
  right: GeneratedImportPair,
): number {
  return compareCodePoints(left.key, right.key);
}

function compareMoves(left: ImportMove, right: ImportMove): number {
  return compareCodePoints(left.oldKey, right.oldKey)
    || compareCodePoints(left.newKey, right.newKey);
}

function compareSuppressions(
  left: ImportMoveSuppression,
  right: ImportMoveSuppression,
): number {
  return compareMoves(left, right)
    || compareCodePoints(left.importId, right.importId)
    || compareCodePoints(left.reason, right.reason);
}

/** Match engine.transform.hcl_string_literal for generated import addresses. */
export function renderHclQuotedString(value: string): string {
  if (typeof value !== "string") {
    return fail(
      "INVALID_HCL_QUOTED_STRING",
      "generated HCL string values must be strings",
    );
  }
  if (value.includes("\0") || !value.isWellFormed()) {
    return fail(
      "INVALID_HCL_QUOTED_STRING",
      "generated HCL string values contain an unsupported character",
    );
  }
  const escaped = value
    .replaceAll("\\", "\\\\")
    .replaceAll("\"", "\\\"")
    .replaceAll("\n", "\\n")
    .replaceAll("\r", "\\r")
    .replaceAll("\t", "\\t")
    // Use a callback because `$` has replacement-pattern semantics in the
    // string overload (`$$` means one literal dollar).
    .replaceAll("${", () => "$${")
    .replaceAll("%{", "%%{");
  return `"${escaped}"`;
}

/** Decode the deliberately small HCL quoted-string grammar emitted above. */
export function parseHclQuotedString(
  text: string,
  start = 0,
): ParsedHclQuotedString {
  if (
    typeof text !== "string"
    ||
    !Number.isSafeInteger(start)
    || start < 0
    || start >= text.length
    || text[start] !== "\""
  ) {
    return fail(
      "INVALID_HCL_QUOTED_STRING",
      "expected a generated HCL quoted string literal",
    );
  }

  const output: string[] = [];
  let index = start + 1;
  while (index < text.length) {
    const character = text[index];
    if (character === "\"") {
      return Object.freeze({ value: output.join(""), end: index + 1 });
    }
    if (character === "\\") {
      index += 1;
      if (index >= text.length) {
        return fail(
          "INVALID_HCL_QUOTED_STRING",
          "generated HCL string has an unterminated escape sequence",
        );
      }
      const escaped = text[index];
      if (escaped === "n") {
        output.push("\n");
      } else if (escaped === "r") {
        output.push("\r");
      } else if (escaped === "t") {
        output.push("\t");
      } else if (escaped === "\"" || escaped === "\\") {
        output.push(escaped);
      } else {
        return fail(
          "INVALID_HCL_QUOTED_STRING",
          "generated HCL string contains an unsupported escape sequence",
        );
      }
      index += 1;
      continue;
    }
    if (text.startsWith("$${", index)) {
      output.push("${");
      index += 3;
      continue;
    }
    if (text.startsWith("%%{", index)) {
      output.push("%{");
      index += 3;
      continue;
    }
    if (character === "\0") {
      return fail(
        "INVALID_HCL_QUOTED_STRING",
        "generated HCL string values cannot contain NUL bytes",
      );
    }
    output.push(character ?? "");
    index += 1;
  }
  return fail(
    "INVALID_HCL_QUOTED_STRING",
    "generated HCL string literal is unterminated",
  );
}

function requirePair(pair: GeneratedImportPair): void {
  if (
    typeof pair !== "object"
    || pair === null
    || Array.isArray(pair)
    || typeof pair.key !== "string"
    || typeof pair.importId !== "string"
  ) {
    fail(
      "INVALID_GENERATED_IMPORTS",
      "generated import keys and ids must be strings",
    );
  }
}

/** Render the byte-canonical import blocks emitted by engine.transform. */
export function renderGeneratedImports(
  resourceType: string,
  pairs: readonly GeneratedImportPair[],
): string {
  requireResourceType(resourceType);
  if (!Array.isArray(pairs)) {
    return fail(
      "INVALID_GENERATED_IMPORTS",
      "generated imports must be an array of key and id pairs",
    );
  }
  if (pairs.length > MAX_GENERATED_IMPORT_PAIRS) {
    return fail(
      "IMPORT_MOVE_LIMIT_EXCEEDED",
      "generated imports exceed the bounded pair contract",
    );
  }
  const keys = new Set<string>();
  for (const pair of pairs) {
    requirePair(pair);
    if (keys.has(pair.key)) {
      return fail(
        "INVALID_GENERATED_IMPORTS",
        "generated imports contain a duplicate Terraform address",
      );
    }
    keys.add(pair.key);
  }
  const blocks = [...pairs].sort(comparePairs).map((pair) => {
    return [
      "import {",
      `  to = module.${resourceType}.${resourceType}.this[${renderHclQuotedString(pair.key)}]`,
      `  id = ${renderHclQuotedString(pair.importId)}`,
      "}",
      "",
    ].join("\n");
  });
  return blocks.join("\n");
}

function expectLiteral(text: string, start: number, literal: string): number {
  if (!text.startsWith(literal, start)) {
    return fail(
      "INVALID_GENERATED_IMPORTS",
      "imports artifact is not a complete canonical generated import file",
    );
  }
  return start + literal.length;
}

/**
 * Parse only complete, byte-canonical Infrawright-generated import files.
 *
 * This intentionally rejects HCL that is semantically equivalent but was not
 * generated by Infrawright. The prior artifact becomes state-move evidence,
 * so comments, partial blocks, repeated addresses, and alternate formatting
 * must not be interpreted heuristically.
 */
export function parseGeneratedImports(
  resourceType: string,
  text: string,
): readonly GeneratedImportPair[] {
  requireResourceType(resourceType);
  if (typeof text !== "string") {
    return fail(
      "INVALID_GENERATED_IMPORTS",
      "generated imports must be canonical text",
    );
  }
  if (text.length === 0) {
    return Object.freeze([]);
  }

  const pairs: GeneratedImportPair[] = [];
  const keys = new Set<string>();
  let cursor = 0;
  while (cursor < text.length) {
    cursor = expectLiteral(
      text,
      cursor,
      `import {\n  to = module.${resourceType}.${resourceType}.this[`,
    );
    const parsedKey = parseHclQuotedString(text, cursor);
    cursor = expectLiteral(text, parsedKey.end, "]\n  id = ");
    const parsedId = parseHclQuotedString(text, cursor);
    cursor = expectLiteral(text, parsedId.end, "\n}\n");

    if (keys.has(parsedKey.value)) {
      return fail(
        "INVALID_GENERATED_IMPORTS",
        "generated imports contain a duplicate Terraform address",
      );
    }
    if (pairs.length >= MAX_GENERATED_IMPORT_PAIRS) {
      return fail(
        "IMPORT_MOVE_LIMIT_EXCEEDED",
        "generated imports exceed the bounded pair contract",
      );
    }
    keys.add(parsedKey.value);
    pairs.push(Object.freeze({
      key: parsedKey.value,
      importId: parsedId.value,
    }));

    if (cursor < text.length) {
      cursor = expectLiteral(text, cursor, "\n");
    }
  }

  if (renderGeneratedImports(resourceType, pairs) !== text) {
    return fail(
      "INVALID_GENERATED_IMPORTS",
      "imports artifact is not byte-canonical generated output",
    );
  }
  return Object.freeze(pairs);
}

function pairsByKey(
  pairs: readonly GeneratedImportPair[],
): ReadonlyMap<string, string> {
  return new Map(pairs.map((pair) => [pair.key, pair.importId]));
}

function keysByImportId(
  pairs: readonly GeneratedImportPair[],
): ReadonlyMap<string, readonly string[]> {
  const grouped = new Map<string, string[]>();
  for (const pair of pairs) {
    const keys = grouped.get(pair.importId);
    if (keys === undefined) {
      grouped.set(pair.importId, [pair.key]);
    } else {
      keys.push(pair.key);
    }
  }
  for (const keys of grouped.values()) {
    keys.sort(compareCodePoints);
  }
  return grouped;
}

function isKeySwap(
  oldKey: string,
  newKey: string,
  oldPairs: ReadonlyMap<string, string>,
  newPairs: ReadonlyMap<string, string>,
): boolean {
  return oldPairs.has(oldKey)
    && oldPairs.has(newKey)
    && newPairs.has(oldKey)
    && newPairs.has(newKey)
    && oldPairs.get(oldKey) === newPairs.get(newKey)
    && oldPairs.get(newKey) === newPairs.get(oldKey);
}

/** Derive only unambiguous, unoccupied state-address moves. */
export function deriveImportMoves(
  resourceType: string,
  oldImportsText: string,
  newImportsText: string,
): ImportMoveDerivation {
  const oldEntries = parseGeneratedImports(resourceType, oldImportsText);
  const newEntries = parseGeneratedImports(resourceType, newImportsText);
  const oldPairs = pairsByKey(oldEntries);
  const newPairs = pairsByKey(newEntries);
  const oldById = keysByImportId(oldEntries);
  const newById = keysByImportId(newEntries);

  const candidates: Array<ImportMove & { readonly importId: string }> = [];
  for (const importId of [...newById.keys()].sort(compareCodePoints)) {
    const oldKeys = oldById.get(importId) ?? [];
    const newKeys = newById.get(importId) ?? [];
    for (const oldKey of oldKeys) {
      for (const newKey of newKeys) {
        if (oldKey !== newKey) {
          if (candidates.length >= MAX_IMPORT_MOVE_CANDIDATES) {
            return fail(
              "IMPORT_MOVE_LIMIT_EXCEEDED",
              "import rename candidates exceed the bounded derivation contract",
            );
          }
          candidates.push({ oldKey, newKey, importId });
        }
      }
    }
  }

  const fromCounts = new Map<string, number>();
  for (const candidate of candidates) {
    fromCounts.set(candidate.oldKey, (fromCounts.get(candidate.oldKey) ?? 0) + 1);
  }

  const moves: ImportMove[] = [];
  const suppressed: ImportMoveSuppression[] = [];
  for (const candidate of candidates) {
    let reason: ImportMoveSuppressionReason | null = null;
    if ((oldById.get(candidate.importId)?.length ?? 0) > 1) {
      reason = "ambiguous";
    } else if ((fromCounts.get(candidate.oldKey) ?? 0) > 1) {
      reason = "duplicate_from";
    } else if (
      isKeySwap(candidate.oldKey, candidate.newKey, oldPairs, newPairs)
    ) {
      reason = "key_swap";
    } else if (
      oldPairs.has(candidate.newKey)
      || newPairs.has(candidate.oldKey)
    ) {
      reason = "destination_occupied";
    }

    if (reason === null) {
      moves.push(Object.freeze({
        oldKey: candidate.oldKey,
        newKey: candidate.newKey,
      }));
    } else {
      suppressed.push(Object.freeze({ ...candidate, reason }));
    }
  }

  moves.sort(compareMoves);
  suppressed.sort(compareSuppressions);
  return Object.freeze({
    moves: Object.freeze(moves),
    suppressed: Object.freeze(suppressed),
  });
}

/** Match engine.transform.render_moves byte-for-byte for derived moves. */
export function renderMovedBlocks(
  resourceType: string,
  moves: readonly ImportMove[],
): string {
  requireResourceType(resourceType);
  if (!Array.isArray(moves)) {
    return fail(
      "INVALID_IMPORT_MOVES",
      "import moves must be an array of address pairs",
    );
  }
  if (moves.length > MAX_IMPORT_MOVE_CANDIDATES) {
    return fail(
      "IMPORT_MOVE_LIMIT_EXCEEDED",
      "import moves exceed the bounded render contract",
    );
  }
  const blocks = moves.map((move) => {
    if (
      typeof move !== "object"
      || move === null
      || Array.isArray(move)
      || typeof move.oldKey !== "string"
      || typeof move.newKey !== "string"
    ) {
      return fail(
        "INVALID_IMPORT_MOVES",
        "move address keys must be strings",
      );
    }
    return [
      "moved {",
      `  from = module.${resourceType}.${resourceType}.this[${renderHclQuotedString(move.oldKey)}]`,
      `  to   = module.${resourceType}.${resourceType}.this[${renderHclQuotedString(move.newKey)}]`,
      "}",
      "",
    ].join("\n");
  });
  return blocks.join("\n");
}

File: go/internal/tfrender/import_moves.go
package tfrender

// Import and moved-block rendering uses a deliberately small ASCII grammar.
// Scanning by UTF-8 byte is safe because every structural token is ASCII and
// non-ASCII key and ID content is copied through unchanged.

import (
	"fmt"
	"regexp"
	"sort"
	"strings"

	"github.com/dvmrry/infrawright-dev/go/internal/canonjson"
	"github.com/dvmrry/infrawright-dev/go/internal/procerr"
)

const (
	maxGeneratedImportPairs = 50_000
	maxImportMoveCandidates = 50_000
)

var importResourceTypePattern = regexp.MustCompile(`^[a-z][a-z0-9_]*$`)

// GeneratedImportPair binds one stable key to its provider import ID.
type GeneratedImportPair struct {
	Key      string
	ImportID string
}

// ImportMove describes a safe Terraform address rename.
type ImportMove struct {
	OldKey string
	NewKey string
}

// ImportMoveSuppressionReason explains why a candidate rename is unsafe.
type ImportMoveSuppressionReason string

const (
	SuppressionAmbiguous           ImportMoveSuppressionReason = "ambiguous"
	SuppressionDuplicateFrom       ImportMoveSuppressionReason = "duplicate_from"
	SuppressionKeySwap             ImportMoveSuppressionReason = "key_swap"
	SuppressionDestinationOccupied ImportMoveSuppressionReason = "destination_occupied"
)

// ImportMoveSuppression is the Go analogue of the ImportMoveSuppression
// interface in the original implementation.
type ImportMoveSuppression struct {
	OldKey   string
	NewKey   string
	ImportID string
	Reason   ImportMoveSuppressionReason
}

// ImportMoveDerivation is the Go analogue of the ImportMoveDerivation
// interface in the original implementation.
type ImportMoveDerivation struct {
	Moves      []ImportMove
	Suppressed []ImportMoveSuppression
}

// ParsedHclQuotedString is the Go analogue of the ParsedHclQuotedString
// interface in the original implementation. End is a byte offset (see
// this file's indexing note above).
type ParsedHclQuotedString struct {
	Value string
	End   int
}

func importMovesFail(code, message string) error {
	return procerr.NewProcessFailure(procerr.NewProcessFailureOptions{
		Code:     code,
		Category: procerr.CategoryDomain,
		Message:  message,
	})
}

func requireImportResourceType(resourceType string) error {
	if !importResourceTypePattern.MatchString(resourceType) {
		return importMovesFail(
			"INVALID_IMPORT_RESOURCE_TYPE",
			"import resource type must be a canonical Terraform identifier",
		)
	}
	return nil
}

// RenderHclQuotedString matches engine.transform.hcl_string_literal for
// generated import addresses. Ports renderHclQuotedString from
// the original implementation.
//
// The TS source's `!value.isWellFormed()` guard (rejecting lone UTF-16
// surrogates) has no Go analogue to check for: a Go string is always valid
// UTF-8 by construction, so there is no lone-surrogate state to reject.
//
// Order matters here exactly as it does in the Node source's chained
// `.replaceAll` calls: backslash/quote/control-char escaping must happen
// before the "${"/"%{" interpolation-guard pass, so that pass only ever
// sees "$"/"{"/"%" characters that were already literal in the input (none
// of the escape replacements above ever introduce a "$", "{", or "%"
// character, so the two passes cannot interact).
func RenderHclQuotedString(value string) (string, error) {
	if strings.ContainsRune(value, 0) {
		return "", importMovesFail(
			"INVALID_HCL_QUOTED_STRING",
			"generated HCL string values contain an unsupported character",
		)
	}
	var body strings.Builder
	for _, r := range value {
		switch r {
		case '\\':
			body.WriteString(`\\`)
		case '"':
			body.WriteString(`\"`)
		case '\n':
			body.WriteString(`\n`)
		case '\r':
			body.WriteString(`\r`)
		case '\t':
			body.WriteString(`\t`)
		default:
			body.WriteRune(r)
		}
	}
	escaped := body.String()
	escaped = strings.ReplaceAll(escaped, "${", "$${")
	escaped = strings.ReplaceAll(escaped, "%{", "%%{")
	return `"` + escaped + `"`, nil
}

// ParseHclQuotedString decodes the deliberately small HCL quoted-string
// grammar RenderHclQuotedString emits, starting at byte offset start (see
// this file's indexing note above). Ports parseHclQuotedString from
// the original implementation.
func ParseHclQuotedString(text string, start int) (ParsedHclQuotedString, error) {
	if start < 0 || start >= len(text) || text[start] != '"' {
		return ParsedHclQuotedString{}, importMovesFail(
			"INVALID_HCL_QUOTED_STRING",
			"expected a generated HCL quoted string literal",
		)
	}
	var output strings.Builder
	index := start + 1
	for index < len(text) {
		switch {
		case text[index] == '"':
			return ParsedHclQuotedString{Value: output.String(), End: index + 1}, nil
		case text[index] == '\\':
			index++
			if index >= len(text) {
				return ParsedHclQuotedString{}, importMovesFail(
					"INVALID_HCL_QUOTED_STRING",
					"generated HCL string has an unterminated escape sequence",
				)
			}
			switch text[index] {
			case 'n':
				output.WriteByte('\n')
			case 'r':
				output.WriteByte('\r')
			case 't':
				output.WriteByte('\t')
			case '"', '\\':
				output.WriteByte(text[index])
			default:
				return ParsedHclQuotedString{}, importMovesFail(
					"INVALID_HCL_QUOTED_STRING",
					"generated HCL string contains an unsupported escape sequence",
				)
			}
			index++
		case strings.HasPrefix(text[index:], "$${"):
			output.WriteString("${")
			index += 3
		case strings.HasPrefix(text[index:], "%%{"):
			output.WriteString("%{")
			index += 3
		case text[index] == 0:
			return ParsedHclQuotedString{}, importMovesFail(
				"INVALID_HCL_QUOTED_STRING",
				"generated HCL string values cannot contain NUL bytes",
			)
		default:
			r, size := decodeRuneAt(text, index)
			output.WriteRune(r)
			index += size
		}
	}
	return ParsedHclQuotedString{}, importMovesFail(
		"INVALID_HCL_QUOTED_STRING",
		"generated HCL string literal is unterminated",
	)
}

// RenderGeneratedImports renders the byte-canonical import blocks emitted
// by engine.transform. Ports renderGeneratedImports from
// the original implementation.
func RenderGeneratedImports(resourceType string, pairs []GeneratedImportPair) (string, error) {
	if err := requireImportResourceType(resourceType); err != nil {
		return "", err
	}
	if len(pairs) > maxGeneratedImportPairs {
		return "", importMovesFail(
			"IMPORT_MOVE_LIMIT_EXCEEDED",
			"generated imports exceed the bounded pair contract",
		)
	}
	seen := make(map[string]bool, len(pairs))
	for _, pair := range pairs {
		if seen[pair.Key] {
			return "", importMovesFail(
				"INVALID_GENERATED_IMPORTS",
				"generated imports contain a duplicate Terraform address",
			)
		}
		seen[pair.Key] = true
	}
	sorted := append([]GeneratedImportPair(nil), pairs...)
	sort.SliceStable(sorted, func(i, j int) bool {
		return canonjson.ComparePythonStrings(sorted[i].Key, sorted[j].Key) < 0
	})
	blocks := make([]string, len(sorted))
	for i, pair := range sorted {
		key, err := RenderHclQuotedString(pair.Key)
		if err != nil {
			return "", err
		}
		id, err := RenderHclQuotedString(pair.ImportID)
		if err != nil {
			return "", err
		}
		blocks[i] = "import {\n" +
			"  to = module." + resourceType + "." + resourceType + ".this[" + key + "]\n" +
			"  id = " + id + "\n" +
			"}\n"
	}
	return strings.Join(blocks, "\n"), nil
}

func expectLiteral(text string, start int, literal string) (int, error) {
	if start > len(text) || !strings.HasPrefix(text[start:], literal) {
		return 0, importMovesFail(
			"INVALID_GENERATED_IMPORTS",
			"imports artifact is not a complete canonical generated import file",
		)
	}
	return start + len(literal), nil
}

// ParseGeneratedImports parses only complete, byte-canonical
// Infrawright-generated import files. Ports parseGeneratedImports from
// the original implementation.
//
// This intentionally rejects HCL that is semantically equivalent but was
// not generated by Infrawright: comments, partial blocks, repeated
// addresses, and alternate formatting must not be interpreted
// heuristically, because the prior artifact becomes state-move evidence.
func ParseGeneratedImports(resourceType, text string) ([]GeneratedImportPair, error) {
	if err := requireImportResourceType(resourceType); err != nil {
		return nil, err
	}
	if len(text) == 0 {
		return []GeneratedImportPair{}, nil
	}

	pairs := []GeneratedImportPair{}
	seen := make(map[string]bool)
	cursor := 0
	prefix := fmt.Sprintf("import {\n  to = module.%s.%s.this[", resourceType, resourceType)
	var err error
	for cursor < len(text) {
		cursor, err = expectLiteral(text, cursor, prefix)
		if err != nil {
			return nil, err
		}
		parsedKey, err := ParseHclQuotedString(text, cursor)
		if err != nil {
			return nil, err
		}
		cursor, err = expectLiteral(text, parsedKey.End, "]\n  id = ")
		if err != nil {
			return nil, err
		}
		parsedID, err := ParseHclQuotedString(text, cursor)
		if err != nil {
			return nil, err
		}
		cursor, err = expectLiteral(text, parsedID.End, "\n}\n")
		if err != nil {
			return nil, err
		}

		if seen[parsedKey.Value] {
			return nil, importMovesFail(
				"INVALID_GENERATED_IMPORTS",
				"generated imports contain a duplicate Terraform address",
			)
		}
		if len(pairs) >= maxGeneratedImportPairs {
			return nil, importMovesFail(
				"IMPORT_MOVE_LIMIT_EXCEEDED",
				"generated imports exceed the bounded pair contract",
			)
		}
		seen[parsedKey.Value] = true
		pairs = append(pairs, GeneratedImportPair{Key: parsedKey.Value, ImportID: parsedID.Value})

		if cursor < len(text) {
			cursor, err = expectLiteral(text, cursor, "\n")
			if err != nil {
				return nil, err
			}
		}
	}

	rendered, err := RenderGeneratedImports(resourceType, pairs)
	if err != nil {
		return nil, err
	}
	if rendered != text {
		return nil, importMovesFail(
			"INVALID_GENERATED_IMPORTS",
			"imports artifact is not byte-canonical generated output",
		)
	}
	return pairs, nil
}

func pairsByKey(pairs []GeneratedImportPair) map[string]string {
	out := make(map[string]string, len(pairs))
	for _, pair := range pairs {
		out[pair.Key] = pair.ImportID
	}
	return out
}

func keysByImportID(pairs []GeneratedImportPair) map[string][]string {
	grouped := make(map[string][]string)
	for _, pair := range pairs {
		grouped[pair.ImportID] = append(grouped[pair.ImportID], pair.Key)
	}
	for id := range grouped {
		keys := grouped[id]
		sort.SliceStable(keys, func(i, j int) bool {
			return canonjson.ComparePythonStrings(keys[i], keys[j]) < 0
		})
		grouped[id] = keys
	}
	return grouped
}

func isKeySwap(oldKey, newKey string, oldPairs, newPairs map[string]string) bool {
	oldOldID, oldOldOK := oldPairs[oldKey]
	oldNewID, oldNewOK := oldPairs[newKey]
	newOldID, newOldOK := newPairs[oldKey]
	newNewID, newNewOK := newPairs[newKey]
	return oldOldOK && oldNewOK && newOldOK && newNewOK &&
		oldOldID == newNewID && oldNewID == newOldID
}

type importMoveCandidate struct {
	oldKey   string
	newKey   string
	importID string
}

// DeriveImportMoves derives only unambiguous, unoccupied state-address
// moves. Ports deriveImportMoves from the original implementation.
func DeriveImportMoves(resourceType, oldImportsText, newImportsText string) (ImportMoveDerivation, error) {
	oldEntries, err := ParseGeneratedImports(resourceType, oldImportsText)
	if err != nil {
		return ImportMoveDerivation{}, err
	}
	newEntries, err := ParseGeneratedImports(resourceType, newImportsText)
	if err != nil {
		return ImportMoveDerivation{}, err
	}
	oldPairs := pairsByKey(oldEntries)
	newPairs := pairsByKey(newEntries)
	oldByID := keysByImportID(oldEntries)
	newByID := keysByImportID(newEntries)

	importIDs := make([]string, 0, len(newByID))
	for id := range newByID {
		importIDs = append(importIDs, id)
	}
	sort.SliceStable(importIDs, func(i, j int) bool {
		return canonjson.ComparePythonStrings(importIDs[i], importIDs[j]) < 0
	})

	var candidates []importMoveCandidate
	for _, importID := range importIDs {
		for _, oldKey := range oldByID[importID] {
			for _, newKey := range newByID[importID] {
				if oldKey != newKey {
					if len(candidates) >= maxImportMoveCandidates {
						return ImportMoveDerivation{}, importMovesFail(
							"IMPORT_MOVE_LIMIT_EXCEEDED",
							"import rename candidates exceed the bounded derivation contract",
						)
					}
					candidates = append(candidates, importMoveCandidate{oldKey, newKey, importID})
				}
			}
		}
	}

	fromCounts := make(map[string]int, len(candidates))
	for _, candidate := range candidates {
		fromCounts[candidate.oldKey]++
	}

	moves := []ImportMove{}
	suppressed := []ImportMoveSuppression{}
	for _, candidate := range candidates {
		var reason ImportMoveSuppressionReason
		switch {
		case len(oldByID[candidate.importID]) > 1:
			reason = SuppressionAmbiguous
		case fromCounts[candidate.oldKey] > 1:
			reason = SuppressionDuplicateFrom
		case isKeySwap(candidate.oldKey, candidate.newKey, oldPairs, newPairs):
			reason = SuppressionKeySwap
		default:
			_, oldHasNewKey := oldPairs[candidate.newKey]
			_, newHasOldKey := newPairs[candidate.oldKey]
			if oldHasNewKey || newHasOldKey {
				reason = SuppressionDestinationOccupied
			}
		}

		if reason == "" {
			moves = append(moves, ImportMove{OldKey: candidate.oldKey, NewKey: candidate.newKey})
		} else {
			suppressed = append(suppressed, ImportMoveSuppression{
				OldKey:   candidate.oldKey,
				NewKey:   candidate.newKey,
				ImportID: candidate.importID,
				Reason:   reason,
			})
		}
	}

	sort.SliceStable(moves, func(i, j int) bool { return compareMoves(moves[i], moves[j]) < 0 })
	sort.SliceStable(suppressed, func(i, j int) bool { return compareSuppressions(suppressed[i], suppressed[j]) < 0 })
	return ImportMoveDerivation{Moves: moves, Suppressed: suppressed}, nil
}

func compareMoves(left, right ImportMove) int {
	if c := canonjson.ComparePythonStrings(left.OldKey, right.OldKey); c != 0 {
		return c
	}
	return canonjson.ComparePythonStrings(left.NewKey, right.NewKey)
}

func compareSuppressions(left, right ImportMoveSuppression) int {
	if c := compareMoves(ImportMove{OldKey: left.OldKey, NewKey: left.NewKey}, ImportMove{OldKey: right.OldKey, NewKey: right.NewKey}); c != 0 {
		return c
	}
	if c := canonjson.ComparePythonStrings(left.ImportID, right.ImportID); c != 0 {
		return c
	}
	return canonjson.ComparePythonStrings(string(left.Reason), string(right.Reason))
}

// RenderMovedBlocks matches engine.transform.render_moves byte-for-byte for
// derived moves. Ports renderMovedBlocks from
// the original implementation.
func RenderMovedBlocks(resourceType string, moves []ImportMove) (string, error) {
	if err := requireImportResourceType(resourceType); err != nil {
		return "", err
	}
	if len(moves) > maxImportMoveCandidates {
		return "", importMovesFail(
			"IMPORT_MOVE_LIMIT_EXCEEDED",
			"import moves exceed the bounded render contract",
		)
	}
	blocks := make([]string, len(moves))
	for i, move := range moves {
		from, err := RenderHclQuotedString(move.OldKey)
		if err != nil {
			return "", err
		}
		to, err := RenderHclQuotedString(move.NewKey)
		if err != nil {
			return "", err
		}
		blocks[i] = "moved {\n" +
			"  from = module." + resourceType + "." + resourceType + ".this[" + from + "]\n" +
			"  to   = module." + resourceType + "." + resourceType + ".this[" + to + "]\n" +
			"}\n"
	}
	return strings.Join(blocks, "\n"), nil
}
